### DISCUSSION TITLE
Passing Xcom values into jinja template parameters

### DISCUSSION THREAD

**Discussion by curious-bug** [apache/airflow#37554] [Q&A]
# Problem
I am currently trying to write a custom task group (similar to whats show in [Astronomer's](https://docs.astronomer.io/learn/task-groups#custom-task-group-classes) documentation).
In my task group I have task_n which takes a value passed into the task group definition and uses it to search for data in SQL.
```python
def tg_definition(value):
    taskn = Operator(
        task_id="task_n",
        sql="SELECT count(*) FROM table WHERE parameter = {{ params.my_value }}",
        params = {"my_value": value}
    )
```

This task group is used in multiple places and sometimes the passed value is a python object:
```python
tg_definition(value = "hello world"):
```
In this case the jinja template renders properly and the query works fine:
```sql -> SELECT count(*) FROM table WHERE parameter = hello world```

However other times the passed value will be an xcom argument:
```python
tg_definition(value = "{{ ti.xcom_pull('external_task') }}")
```

In this case, the xcom argument does not resolve and the query renders improperly:
```sql -> SELECT count(*) FROM table WHERE parameter = {{ ti.xcom_pull('external_task') }}```

I understand that this is expected behavior since Jinja does not process nested templates, however I was hoping that airflow implemented some sort of workaround.

# My current solution
Currently, to allow me to use the passed value in the jinja parameters, I have created another task `template_param` that takes `value`  and returns a dictionary with the object imbedded in it
```python
def tg_definition(value):
    @task
    def template_param(val):
        return [{"my_value": val}]

    taskn = Operator.partial(
        task_id="task_n",
        sql="SELECT count(*) FROM table WHERE parameter = {{ params.my_value }}",
    ).expand(params=template_param(val = value) )
```
Since any xcom templates are rendered in the python task this works when `value` is an xcom or a normal python object.
However along with needing an additional task, the param dictionary must be given to `task_n` using a mapped task (for some reason I dont understand).

Overall this solution is somewhat clunky. I would like to be able to pass my desired values straight into the template params without needing the intermediary task.
Any help is much apricated.

**Comment by potiuk** [comment]
You have not explained in which cases you get templates and in which values. That could be helpful to know in order to help you, but from what I see the behaviour is pretty expected. 

When you pass a value to operator, it will get JINJA templates, this is expected and normal and this is how they are passed to Operators, this is inherent part of the operator's definition. Those templates are rendered automatically when the field is marked as templated (i.e. when you have in the constructor:

```python
def __init__(a_field):
self.a_field = a_field
```

When the field is set with JINJA template, it will be stored in the `a_field` as Jinja template, and it will be rendered with the context at runtime, when the `execute` method of the operator is called. More about it in https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/operators.html#jinja-templating . So maybe your problem is that you are accessung (and expecting the value to be different) before the execute method is run during DAG execution. If so, then you have to make sure you only access those field in the flow of the execute method.

But I am largely guessing now, maybe you can explain better when you get the values with and when without template. For now I think that "with" is before the execute and "without" is when execute is running. But again - maybe I misunderstood your problem and you need to explain it better.

**Reply by curious-bug** [reply]
The reason I can get both template values and pure python objects is just down to how the task group is implemented in a particular dag.
Some will generate the value while others have the value hardcoded. Though the majority will be templated.

As I said, I understand this is expected behavior, it's just an effect of how jinja templating is implemented in airflow.
The root of my question is if there is a way to enforce templating on values in the `params` argument of a given task.

**Reply by potiuk** [reply]
I don't think you need to use params at all. I think this should do what you want (assuming sql is templated):

```python
def tg_definition(value):
    taskn = Operator(
        task_id="task_n",
        sql=f"SELECT count(*) FROM table WHERE parameter = \"{value}\"",
    )
```

**Reply by curious-bug** [reply]
I tried that and it does work but I personally find it somewhat clunky
Though I may end up returning to it.

**Reply by potiuk** [reply]
Well. The other option you have is to add a feature of rendering Jinja template in Params. I think there are several people who raised this in the past (you can search through issues and discussions) and have in someone who would have incentive to do it and join the 2800+ contributors to airflow whow would do it is a good idea. Many features of Airflow were created because a user wanted something and contributed it back - this is also a great way to give back to the community for the free software you use.

I heartily encourage you to do so.
